### PR TITLE
Add keyboard input to layer surfaces

### PR DESCRIPTION
**Pull request by SirCmpwn**
Per the spec, the topmost layer surface above the shell surface layer always has keyboard focus. Beneath the shell surface layer, layer surfaces will receive keyboard focus normally (when clicked). Note that layer surfaces beneath the shell surface layer receive a sort of temporary focus from the seat - they do not participate in the usual focus list for alt-tabbing and such. Layer surfaces that do not want keyboard input continue to not receive keyboard enter events or take focus away from other surfaces when clicked.

To request keyboard events, pass `-k` to the example layer shell and look for key events in the log.

### CODE REVIEW

File: rootston/cursor.c [hunk 1]
```diff
@@ -271,6 +271,12 @@ static void roots_cursor_press_button(struct roots_cursor *cursor,
 			break;
 		case WLR_BUTTON_PRESSED:
 			roots_seat_set_focus(seat, view);
+			if (surface && !view) {
+				struct wlr_layer_surface *layer = surface->role_data;
```
Comment (emersion) [inline rootston/cursor.c]: This is very meh™. We want to get rid of `role_data`, and it's an internal field.

Comment (SirCmpwn) [reply rootston/cursor.c]: We need some means of getting a wlr_layer_surface/wlr_xdg_surface/etc from a wlr_surface, be it this or some other way

File: rootston/layer_shell.c [hunk 2]
```diff
@@ -192,6 +192,35 @@ void arrange_layers(struct roots_output *output) {
...
 			&usable_area, false);
+
+	// Find topmost keyboard interactive layer, if such a layer exists
+	uint32_t layers_above_shell[] = {
+		ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY,
+		ZWLR_LAYER_SHELL_V1_LAYER_TOP,
+	};
+	struct roots_layer_surface *layer, *topmost = NULL;
+	for (size_t i = 0;
+			i < sizeof(layers_above_shell) / sizeof(layers_above_shell[0]);
```
Comment (emersion) [inline rootston/layer_shell.c]: Create a temporary variable for the size?

File: rootston/cursor.c (hunk 1 above)
Comment (emersion) [reply rootston/cursor.c]: Just iterate over the layer surfaces list.

Comment (SirCmpwn) [reply rootston/cursor.c]: That is Very Meh™.

Comment (emersion) [reply rootston/cursor.c]: https://github.com/swaywm/wlroots/issues/527

Find another way. Set `wlr_surface.data` for instance.

Comment (SirCmpwn) [reply rootston/cursor.c]: Meh.

File: rootston/seat.c [hunk 3]
```diff
@@ -800,6 +804,37 @@ void roots_seat_set_focus(struct roots_seat *seat, struct roots_view *view) {
...
+ * You also cannot alt-tab between layer surfaces and shell surfaces.
+ */
+void roots_seat_set_focus_layer(struct roots_seat *seat,
+		struct wlr_layer_surface *layer) {
+	struct wlr_keyboard *keyboard = wlr_seat_get_keyboard(seat->seat);
+	if (!layer) {
+		seat->focused_layer = NULL;
+		return;
+	}
+	if (seat->has_focus) {
```
Comment (emersion) [inline rootston/seat.c]: What happens if there's a fullscreen view? What happens if the view is focused by another client?

Calling `roots_seat_set_focus(seat, NULL)` could work better.

Comment (SirCmpwn) [reply rootston/seat.c]: So what if there's a fullscreen view? You can't click layers beneath the shell layer if there is. Layers above the shell layer behave as expected.

File: rootston/cursor.c (hunk 1 above)
Comment (SirCmpwn) [reply rootston/cursor.c]: We need two functions:

`wlr_layer_surface_from_wlr_surface`

`wlr_surface_is_layer_surface`

Same for xwayland, xdg-shell, etc. This is out of scope for this PR, come back and fix this later.

File: include/wlr/types/wlr_wl_shell.h [hunk 4]
```diff
@@ -150,4 +150,9 @@ struct wlr_wl_shell_surface *wlr_wl_shell_surface_popup_at(
 		struct wlr_wl_shell_surface *surface, double sx, double sy,
 		double *popup_sx, double *popup_sy);
 
+bool wlr_surface_is_wl_surface(struct wlr_surface *surface);
```
Comment (emersion) [inline include/wlr/types/wlr_wl_shell.h]: wlr_surface_is_wl_shell_surface

Comment (SirCmpwn) [reply include/wlr/types/wlr_wl_shell.h]: derp

**Pull request merged**